src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs

```diff
@@ -250,6 +264,42 @@ internal static bool TryCreate(SyntheticBoundNodeFactory F, MethodSymbol method,
             throw ExceptionUtilities.UnexpectedValue(method);
         }
 
+        private static NamedTypeSymbol ValidateBuilderType(SyntheticBoundNodeFactory F, object builderAttributeArgument, Accessibility desiredAccessibility, int desiredArity)
+        {
+            var builderType = builderAttributeArgument as NamedTypeSymbol;
+
+            bool isArityOk;
+            if (desiredArity == 0)
+            {
+                isArityOk = builderType?.IsGenericType == false;
+            }
+            else if (desiredArity == 1)
+            {
+                isArityOk = builderType?.IsUnboundGenericType == true &&
+                            builderType?.ContainingType?.IsGenericType != true;
+            }
+            else
+            {
+                isArityOk = false;
+                Debug.Assert(false, "expected arity 0 or 1");
+            }
```

Comment:
Fixed in https://github.com/dotnet/roslyn/pull/13405/commits/f2cdba83c83d34ddac4875def41674d352fc5b7e#diff-995d1bf33206eef8c980dbd08b8601b0R267 - I changed to bool, and streamlined the number of null checks.